csds-client/client/client.go

```diff
@@ -0,0 +1,245 @@
+	requestFilePtr := flag.String("request_file", "", "yaml file that defines the csds request")
+	requestYamlPtr := flag.String("request_yaml", "", "yaml string that defines the csds request")
+	jwtPtr := flag.String("jwt_file", "", "path of the -jwt_file")
+	configFilePtr := flag.String("file_to_save_config", "", "the file name to save config")
+	monitorFreqPtr := flag.Duration("monitor_freq", 0, "the frequency of sending request in monitor mode (e.g. 500ms, 2s, 1m ...)")
+
+	flag.Parse()
+
+	f := Flag{
+		uri:         *uriPtr,
+		platform:    *platformPtr,
+		authnMode:   *authnModePtr,
+		apiVersion:  *apiVersionPtr,
+		requestFile: *requestFilePtr,
+		requestYaml: *requestYamlPtr,
+		jwt:         *jwtPtr,
+		configFile:  *configFilePtr,
+		monitorFreq: *monitorFreqPtr,
+	}
+
+	return f
+}
+
+// parseNodeMatcher parses the csds request yaml to nodematcher
+func (c *Client) parseNodeMatcher() error {
+	if c.info.requestFile == "" && c.info.requestYaml == "" {
+		return errors.New("missing request yaml")
+	}
+
+	var nodematchers []*envoy_type_matcher.NodeMatcher
+	if err := parseYaml(c.info.requestFile, c.info.requestYaml, &nodematchers); err != nil {
+		return err
+	}
+
+	c.nm = nodematchers
+
+	// check if required fields exist in nodematcher
+	switch c.info.platform {
+	case "gcp":
+		switch c.info.uri {
+		case "trafficdirector.googleapis.com:443":
+			keys := []string{"TRAFFICDIRECTOR_GCP_PROJECT_NUMBER", "TRAFFICDIRECTOR_NETWORK_NAME"}
+			for _, key := range keys {
+				if value := getValueByKeyFromNodeMatcher(c.nm, key); value == "" {
+					return fmt.Errorf("missing field %v in NodeMatcher", key)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// connWithAuth connects to uri with authentication
+func (c *Client) connWithAuth() error {
+	var scope string
+	switch c.info.authnMode {
+	case "jwt":
+		if c.info.jwt == "" {
+			return errors.New("missing jwt file")
+		}
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewServiceAccountFromFile(c.info.jwt, scope)
+			if err != nil {
+				return err
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return nil
+		}
+	case "auto":
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewApplicationDefault(context.Background(), scope) // Application Default Credentials (ADC)
+			if err != nil {
+				return err
+			}
+
+			// parse GCP project number as header for authentication
+			var key string
+			switch c.info.uri {
+			case "trafficdirector.googleapis.com:443":
```

Comment:
maybe condition on "gcp" instead of td uri?

Reply:
There's a condition on "gcp" before this. 
If the `platform` is set to gcp, then the client will parse GCP project number for authentication. Then it will check the `uri`. For here, the condition on `uri` is used to get the key of project number based on different apis on gcp. For example, for TD, the key is `TRAFFICDIRECTOR_GCP_PROJECT_NUMBER`. For other apis on gcp, it may be something different.

Reply:
Since you said that it's safe to assume (at least for now) gcp and td are 1:1, I've deleted the check for uri after check for gcp.